beacon-chain/sync/pending_attestations_queue.go

```diff
@@ -91,86 +91,127 @@ func (s *Service) processPendingAtts(ctx context.Context) error {
+	// validation steps.
+	valRes, err := s.validateAggregatedAtt(ctx, att)
+	if err != nil {
+		log.WithError(err).Debug("Pending aggregated attestation failed validation")
+	}
+	aggValid := pubsub.ValidationAccept == valRes
+	if s.validateBlockInAttestation(ctx, att) && aggValid {
+		if features.Get().EnableExperimentalAttestationPool {
+			if err = s.cfg.attestationCache.Add(aggregate); err != nil {
+				log.WithError(err).Debug("Could not save aggregate attestation")
+				return
 			}
-			preState, err := s.cfg.chain.AttestationTargetState(ctx, data.Target)
-			if err != nil {
-				log.WithError(err).Debug("Could not retrieve attestation prestate")
-				continue
+		} else {
+			if err := s.cfg.attPool.SaveAggregatedAttestation(aggregate); err != nil {
+				log.WithError(err).Debug("Could not save aggregate attestation")
+				return
 			}
+		}
+
+		s.setAggregatorIndexEpochSeen(aggregate.GetData().Target.Epoch, att.AggregateAttestationAndProof().GetAggregatorIndex())
 
-			valid, err := s.validateUnaggregatedAttWithState(ctx, aggregate, preState)
-			if err != nil {
-				log.WithError(err).Debug("Pending unaggregated attestation failed validation")
-				continue
+		// Broadcasting the signed attestation again once a node is able to process it.
+		if err := s.cfg.p2p.Broadcast(ctx, att); err != nil {
+			log.WithError(err).Debug("Could not broadcast")
+		}
+	}
+}
+
+func (s *Service) processUnaggregated(ctx context.Context, att ethpb.Att) {
+	data := att.GetData()
+
+	// This is an important validation before retrieving attestation pre state to defend against
+	// attestation's target intentionally reference checkpoint that's long ago.
+	// Verify current finalized checkpoint is an ancestor of the block defined by the attestation's beacon block root.
+	if !s.cfg.chain.InForkchoice(bytesutil.ToBytes32(data.BeaconBlockRoot)) {
+		log.WithError(blockchain.ErrNotDescendantOfFinalized).Debug("Could not verify finalized consistency")
+		return
+	}
+	if err := s.cfg.chain.VerifyLmdFfgConsistency(ctx, att); err != nil {
+		log.WithError(err).Debug("Could not verify FFG consistency")
+		return
+	}
+	preState, err := s.cfg.chain.AttestationTargetState(ctx, data.Target)
+	if err != nil {
+		log.WithError(err).Debug("Could not retrieve attestation prestate")
+		return
+	}
+	committee, err := helpers.BeaconCommitteeFromState(ctx, preState, data.Slot, att.GetCommitteeIndex())
+	if err != nil {
+		log.WithError(err).Debug("Could not retrieve committee from state")
+		return
+	}
+	valid, err := validateAttesterData(ctx, att, committee)
+	if err != nil {
+		log.WithError(err).Debug("Could not validate attester data")
+		return
+	} else if valid != pubsub.ValidationAccept {
+		log.Debug("Attestation failed attester data validation")
+		return
+	}
+
+	var singleAtt *ethpb.SingleAttestation
+	if att.Version() >= version.Electra {
+		var ok bool
+		singleAtt, ok = att.(*ethpb.SingleAttestation)
+		if !ok {
```

Comment:
I have a question on this path. According to the Electra spec `AggregateAndProof` [here](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/validator.md#aggregateandproof) contains an `Attestation` object inside which is never a `SingleAttestation` according to [this definition](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#attestation) in the Electra spec